[RISCV] Add MIPS extensions

djtodoro opened the pull request:
Adding two extensions for MIPS p8700 CPU:
  1. cmove (conditional move)
  2. lsp (load/store pair)

The official product page here:
https://mips.com/products/hardware/p8700

djtodoro committed:
[RISCV] Add MIPS extensions

Adding two extensions for MIPS p8700 CPU:
  1. cmove (conditional move)
  2. lsp (load/store pair)

The official product page here:
https://mips.com/products/hardware/p8700

topperc commented:
This still isn't broken down enough. We usually like to see assembler support in separate patches from code generation.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -0,0 +1,82 @@
+let Predicates = [HasVendorMIPSCMove], hasSideEffects = 0, mayLoad = 0, mayStore = 0, DecoderNamespace = "Xmipscomve" in {
+def CCMOV : RVInstR4<0b11, 0b011, OPC_CUSTOM_0, (outs GPR:$rd),
+                    (ins GPR:$rs1, GPR:$rs2, GPR:$rs3),
+                    "ccmov", "$rd, $rs2, $rs1, $rs3">,
```
Vendor instructions should have a prefix like `mips.` to avoid collisions with other vendors or future standard instruction names.

jrtc27 replied:
For upstreaming such a prefix must be registered: https://github.com/riscv-non-isa/riscv-toolchain-conventions/blob/08ae8fe66141c12ba436678c60a05da9ac08c4e6/src/toolchain-conventions.adoc#assembly-mnemonic

djtodoro replied:
Yep, I have added it. The PR for `mips.` prefix is here https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/69.

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1395,6 +1395,19 @@ def NoConditionalMoveFusion  : Predicate<"!Subtarget->hasConditionalMoveFusion()
 def TuneMIPSP8700
     : SubtargetFeature<"mips-p8700", "RISCVProcFamily", "MIPSP8700",
                        "MIPS p8700 processor">;
+def FeatureVendorMIPSCMove : SubtargetFeature<"xmipscmove", "HasVendorMIPSCMove",
```
FeatureVendor* should be in the vendor extension section of this file not the tuning section.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrFormats.td:
```diff
@@ -514,6 +514,78 @@ class RVInstJ<RISCVOpcode opcode, dag outs, dag ins, string opcodestr,
+  let Inst{19-15} = rs1;
+  let Inst{14-12} = 0b100;
+  let Inst{11-7} = rd1;
+  let Inst{6-0} = 0b0001011;
```
This use the `OPC_CUSTOM_0.Value` for readability

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,370 @@
+  AA = &getAnalysis<AAResultsWrapperPass>().getAAResults();
+  ModifiedRegUnits.init(*TRI);
+  UsedRegUnits.init(*TRI);
+  UseLoadStorePair = Subtarget.useLoadStorePairs();
```
This is already checked to be true on line 82. Why do we need a member variable for it?

djtodoro replied:
No need any more, thanks

topperc commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -62,6 +62,15 @@ static cl::opt<unsigned> RISCVMinimumJumpTableEntries(
     cl::desc("Set minimum number of entries to use a jump table on RISCV"));
 
+static cl::opt<bool>
+    UseLoadStorePairsOpt("riscv-load-store-pairs",
```
Why is this off by default?

djtodoro replied:
Well, we have not seen benefits on some important applications/benchmarks for us when using this extension with `p8700` CPU, so that is why we keep it `OFF` by default for now.

jrtc27 replied:
I worry that this name is going to get confused with Zilsd

djtodoro replied:
Hmm, you are right. I will add `mips-` to remove confusion. Thanks @jrtc27!

topperc commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -389,6 +390,13 @@ class RISCVPassConfig : public TargetPassConfig {
     }
+
+    const RISCVSubtarget &ST = C->MF->getSubtarget<RISCVSubtarget>();
+    if (!ST.getMacroFusions().empty() && ST.useLoadStorePairs()) {
```
This needs to be integrated with the `EnablePostMISchedLoadStoreClustering` on line 386. Otherwise we will have multiple load/store cluster mutations.

djtodoro replied:
Hm, I agree

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.td:
```diff
@@ -257,6 +257,146 @@ def simm12 : RISCVSImmLeafOp<12> {
   }];
 }
 
+// A 7-bit unsigned immediate where the least significant two bits are zero.
```
Why do these need to move? RISCVInstrInfoXMips.td is included after RISCVInstrInfoC.td

djtodoro replied:
No need any more, thanks

topperc commented:
Missing tests in test/MC/RISCV for the assembler and disassembler

lenary commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4963,6 +4963,10 @@ def msave_restore : Flag<["-"], "msave-restore">, Group<m_riscv_Features_Group>,
 def mno_save_restore : Flag<["-"], "mno-save-restore">, Group<m_riscv_Features_Group>,
   HelpText<"Disable using library calls for save and restore">;
 } // let Flags = [TargetSpecific]
+def mload_store_pairs : Flag<["-"], "mload-store-pairs">, Group<m_riscv_Features_Group>;
+def mno_load_store_pairs : Flag<["-"], "mno-load-store-pairs">, Group<m_riscv_Features_Group>;
+def mccmov : Flag<["-"], "mccmov">, Group<m_riscv_Features_Group>;
+def mno_ccmov : Flag<["-"], "mno-ccmov">, Group<m_riscv_Features_Group>;
```
The RISC-V backend has not added additional options (i.e. those not in the list `-march=`/`-mcpu=`/`-mabi=`) that change the target architecture extensions, and my understanding is this is on purpose (unlike in other targets). I think these would need more discussion before they are added.

I could understand gating optimisations to introduce these instructions behind additional arguments, but these seem to control whether specific instructions are or are not legal.

djtodoro replied:
Well, you are right. We have used those for testing purposes, and I will remove it. Thank you for the point.

djtodoro committed:
Addressing comments

djtodoro committed:
Remove RISCVLoadStoreOptimizer

This will be added in a separate commit.

djtodoro commented:
@topperc Thanks for the comments!

> This still isn't broken down enough. We usually like to see assembler support in separate patches from code generation.

I have removed `RISCVLoadStoreOptimizer` Pass, and will add it in a separate PR/commit.

> Missing tests in test/MC/RISCV for the assembler and disassembler

Added.

lenary commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
Please can you undo these whitespace changes, given you're not really making changes to this file

lenary replied:
You still have a whitespace change to this file?

lenary reviewed (commented):
Some small notes, the one about using RISCVExtension will probably have the most knock-on work (run both the llvm and the clang test suites after that change, to see where that info gets to)

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -0,0 +1,82 @@
+let Predicates = [HasVendorMIPSLoadStorePairs], hasSideEffects = 0, DecoderNamespace = "Xmipslsp" in {
+def LWP : LWPFormat<(outs GPR:$rd1, GPR:$rd2), (ins GPR:$rs1, uimm7_lsb00:$imm7),
+                    "mips.lwp", "$rd1, $rd2, ${imm7}(${rs1})">,
+                    Sched<[WriteLDW, WriteLDW, ReadMemBase]> {
```
`Sched` should be aligned to the same column as `LWPFormat`

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -0,0 +1,82 @@
+// MIPS extensions
+//===----------------------------------------------------------------------===//
+
+let Predicates = [HasVendorMIPSCMove], hasSideEffects = 0, mayLoad = 0, mayStore = 0, DecoderNamespace = "Xmipscomve" in {
```
Xmipscomve -> Xmipscmove?

djtodoro replied:
thanks!

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1448,3 +1448,18 @@ def FeatureTaggedGlobals : SubtargetFeature<"tagged-globals",
     "true", "Use an instruction sequence for taking the address of a global "
     "that allows a memory tag in the upper address bits">;
+
+def FeatureVendorMIPSCMove : SubtargetFeature<"xmipscmove", "HasVendorMIPSCMove",
```
FeatureVendorMIPSCMove -> FeatureVendorXMIPSCMove so it matches the name of the extension

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1448,3 +1448,18 @@ def FeatureTaggedGlobals : SubtargetFeature<"tagged-globals",
+    : SubtargetFeature<"xmipslsp", "HasMIPSLSP", "true",
+                       "Optimize for hardware load-store bonding">;
+def HasVendorMIPSLoadStorePairs
+    : Predicate<"Subtarget->useLoadStorePairs()">,
```
There's no patterns for load/store pair in tablegen so there is no reason this can't be "Subtarget->hasVendorMIPSCMove()"

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1448,3 +1448,18 @@ def FeatureTaggedGlobals : SubtargetFeature<"tagged-globals",
+def FeatureVendorMIPSCMove : SubtargetFeature<"xmipscmove", "HasVendorMIPSCMove",
+                                       "true", "Using CCMov",
+                                       [Feature64Bit]>;
+def HasVendorMIPSCMove
```
I think I would rather see this split into `HasVendorMIPSCMove` for assembler, and `UseCCMovInsn` for isel patterns.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoC.td:
```diff
@@ -109,6 +109,21 @@ def uimm7_lsb00 : RISCVOp,
   }];
 }
 
+// A 7-bit unsigned immediate where the least significant three bits are zero.
```
This should be defned in Mips specific file if that's the only place it is used.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -0,0 +1,82 @@
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+
```
Extra blank line?

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoXMips.td:
```diff
@@ -0,0 +1,82 @@
+def : Pat<(select (XLenVT (seteq (XLenVT GPR:$x), (XLenVT GPR:$y))),
+                  (XLenVT GPR:$rs3), (XLenVT GPR:$rs1)),
+          (CCMOV GPR:$rs1, (XOR GPR:$x, GPR:$y), GPR:$rs3)>;
+def : Pat<(select (XLenVT (setuge (XLenVT GPR:$x), (XLenVT GPR:$y))),
```
I don't think setuge, setge, setle, setule will make it to isel so any patterns that use them can't be tested.

djtodoro replied:
I see, it makes sense! :D

Initially, I created it against an older LLVM, so I have missed this...Thanks!

djtodoro commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1448,3 +1448,18 @@ def FeatureTaggedGlobals : SubtargetFeature<"tagged-globals",
     "true", "Use an instruction sequence for taking the address of a global "
     "that allows a memory tag in the upper address bits">;
+
+def FeatureVendorMIPSCMove : SubtargetFeature<"xmipscmove", "HasVendorMIPSCMove",
+                                       "true", "Using CCMov",
+                                       [Feature64Bit]>;
```
>It would be better if this and def FeatureVendorMIPSLoadStorePairs both used RISCVExtension. I think this means some changes to the functions on riscvsubtarget too, it should be clear to follow from the other vendor extensions. This will also have a knock-on effect on some other tests, especially relating to RISCVISAInfo (and also some in clang iirc), but this is for the better - it means the compiler can show it supports your extensions 😄

Good point. Yes, I have added it, and resolved test failures (there is one unit test as well). Thanks :)

>Please may you also put these definitions beside the other vendor extensions, rather than at the end with these codegen options? The section starts somewhere around line 1020, and you'll see each vendor's extensions are with each other.

Sure.

djtodoro committed:
Addressing comments

djtodoro commented on llvm/lib/Target/RISCV/RISCVInstrFormats.td:
```diff
@@ -514,6 +514,78 @@ class RVInstJ<RISCVOpcode opcode, dag outs, dag ins, string opcodestr,
+  let Inst{19-15} = rs1;
+  let Inst{14-12} = 0b101;
+  let Inst{11-9} = imm7{4-2};
+  let Inst{8-0} = 0b010001011;
```
@topperc you meant `let Inst{6-0} = OPC_CUSTOM_0.Value;` I think :)

topperc replied:
> @topperc you meant `let Inst{6-0} = OPC_CUSTOM_0.Value;` I think :)

That's what I meant. Thanks.

djtodoro committed:
clang-format

lenary commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -238,3 +247,11 @@ void RISCVSubtarget::overridePostRASchedPolicy(MachineSchedPolicy &Policy,
   }
 }
+
+bool RISCVSubtarget::useLoadStorePairs() const {
+  return UseMIPSLoadStorePairsOpt && HasVendorXMIPSLSP;
+}
```
Nit: I'm not sure this is used any more.

lenary reviewed (approved):
Two tiny issues, but I think I'm happy for this to land.

djtodoro committed:
Remove unused code

djtodoro commented:
@lenary thanks!

>You still have a whitespace change to this file?

No, the whitespaces are there on `main` branch. I applied `clang-format` initially, and that is why it was included in one of my previous patches.

djtodoro committed:
fixup: Move inst formats to mips td file

djolertrk commented on llvm/lib/Target/RISCV/RISCVInstrFormats.td:
```diff
@@ -514,6 +514,80 @@ class RVInstJ<RISCVOpcode opcode, dag outs, dag ins, string opcodestr,
   let Inst{6-0} = opcode.Value;
 }
 
+//===----------------------------------------------------------------------===//
```
Makes sense. Thanks.

djtodoro replied:
It makes sense. Thanks.

djolertrk commented:
@topperc looks ok now?

djtodoro commented:
@topperc Is it ok now? :)

topperc commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1228,6 +1228,21 @@ def HasVendorXCVbi
+def FeatureVendorXMIPSLSP
+    : RISCVExtension<1, 0, "MIPS optimization for hardware load-store bonding">;
+def HasVendorXMIPSLSP
+    : Predicate<"Subtarget->hasVendorXMIPSCMove()">,
```
hasVendorXMIPSCMove -> hasVendorXMIPSLSP?

djtodoro replied:
IIUC, I think you proposed we use `hasVendorXMIPSCMove` for both.

djtodoro replied:
https://github.com/llvm/llvm-project/pull/121394#discussion_r1918794675

topperc replied:
I think I meant to write `hasVendorXMIPSLSP()`. I think I didn't want it to be `useLoadStorePairs()`.

djtodoro replied:
Aha ok. I understand.

djtodoro committed:
fixup: Use hasVendorXMIPSLSP for l/d pairs

topperc reviewed (approved):
LGTM
